### PR TITLE
add broadcasting support for min/max/sum/mean

### PR DESCRIPTION
**Pull request by linkerzhang**

### CODE REVIEW

File: onnx/defs/math/defs.cc [hunk 1]
```diff
@@ -493,82 +493,50 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+        "data_0",
+        "List of tensors for " + std::string(name) + ".",
+        "T",
+        OpSchema::Variadic);
+    schema.Output(0, name, "Output tensor.", "T");
+    schema.TypeConstraint(
+        "T",
+        {"tensor(float16)", "tensor(float)", "tensor(double)"},
+        "Constrain input and output types to float tensors.");
+    schema.TypeAndShapeInferenceFunction(propagateShapeAndTypeFromFirstInput);
```
[the remark this answers is not in the document]
Comment (linkerzhang) [reply onnx/defs/math/defs.cc]: Yep. I'm fixing it.

[the remark this answers is not in the document]
Comment (linkerzhang) [reply onnx/defs/math/defs.cc]: Fixed.

File: onnx/defs/math/defs.cc [hunk 2]
```diff
@@ -493,82 +493,50 @@ ONNX_OPERATOR_SET_SCHEMA(
             "Constrain input and output types to float tensors.")
         .TypeAndShapeInferenceFunction(propagateShapeAndTypeFromFirstInput));
 
-static const char* Max_ver6_doc = R"DOC(
-Element-wise max of each of the input tensors. All inputs and outputs must
-have the same shape and data type.
+std::function<void(OpSchema&)> ElementwiseOpDocGenerator(const char* name) {
```
[the remark this answers is not in the document]
Comment (linkerzhang) [reply onnx/defs/math/defs.cc]: done.

File: onnx/defs/math/defs.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (houseroad) [reply onnx/defs/math/defs.cc]: Delete this line?

[the remark this answers is not in the document]
Comment (linkerzhang) [reply onnx/defs/math/defs.cc]: Great catch! Thank you! :)

File: onnx/defs/shape_inference.h [hunk 3]
```diff
@@ -291,85 +309,96 @@ inline void updateOutputShape(
...
+        }
+        ++num_symbolic_dims;
       }
     }
+
+    if (dim_value != 1 || num_symbolic_dims == 0) {
+      resultShape.add_dim()->set_dim_value(dim_value);
+    } else if (num_symbolic_dims == 1) {
+      *resultShape.add_dim() = symbolic_dim;
+	}
```
Comment (houseroad) [inline onnx/defs/shape_inference.h]: indent here

File: onnx/defs/shape_inference.h [hunk 4]
```diff
@@ -291,85 +309,96 @@ inline void updateOutputShape(
...
-      }
-    } else if (shapeL.dim(il).has_dim_value()) {
-      auto dimL = shapeL.dim(il).dim_value();
-      if (dimL == 1) {
-        *newdim = shapeR.dim(ir);
-      } else {
-        newdim->set_dim_value(dimL);
+  for (int i = 0; i < result_shape_size; ++i) {
+    int64_t dim_value = 1;
+    TensorShapeProto_Dimension symbolic_dim;
```
Comment (houseroad) [inline onnx/defs/shape_inference.h]: What's the purpose of symbolic_dim? The logic seems incomplete.

Comment (linkerzhang) [reply onnx/defs/shape_inference.h]: offline explained, symbolic dim is used,
1, if there's one symbolic dim and all others are 1, then symbolic dim will be used in result shape.
2. if there're symbolic dims and one dim value > 1, then the result shape will take the >1 value.
3. if there're more than 1 symbolic dim and all others are 1, then the result shape will not be set.

**Pull request merged**